### PR TITLE
Add bitmasked-option for tbr analysis

### PR DESCRIPTION
**Pull request by vaithak**

### CODE REVIEW

File: tools/ClangPlugin.cpp [hunk 1]
```diff
@@ -318,6 +315,19 @@ namespace clad {
       m_HasRuntime = !R.empty();
       return m_HasRuntime;
     }
+
+    void CladPlugin::SetRequestOptions(RequestOptions& opts) const {
+      /// ---------- Set TBR analysis flag ----------
```
Comment (vgvassilev) [inline tools/ClangPlugin.cpp:320]: Would it help to increase readability to move the setting of each flag into a separate file static routine?

**Review by vgvassilev — COMMENTED** [review]
This LGTM, but I'd wait for @PetroZarytskyi's review.

**Comment by vgvassilev** [comment]
I am wondering if we could draw some inspiration from LLVM's BitmaskEnum?

**Comment by vaithak** [comment]
> I am wondering if we could draw some inspiration from LLVM's BitmaskEnum?

I went through their code and read about the [BitmaskType in C++](https://en.cppreference.com/w/cpp/named_req/BitmaskType)? 
To confirm, using this approach, we will have 2 benefits:
- Setting of bit only possible through `clad::opts::...` enum and not through some other unrelated enums or integer literals.
- This can help in a single bit for tbr analysis? Disabling using `~clad::opts::tbr_analysis` (no separate enable or disable).

Is this correct? If so, I think yeah, we should have an interface using this approach.

**Comment by PetroZarytskyi** [comment]
> > I am wondering if we could draw some inspiration from LLVM's BitmaskEnum?
> 
> I went through their code and read about the [BitmaskType in C++](https://en.cppreference.com/w/cpp/named_req/BitmaskType)? To confirm, using this approach, we will have 2 benefits:
> 
> * Setting of bit only possible through `clad::opts::...` enum and not through some other unrelated enums or integer literals.
> * This can help in a single bit for tbr analysis? Disabling using `~clad::opts::tbr_analysis` (no separate enable or disable).
> 
> Is this correct? If so, I think yeah, we should have an interface using this approach.

Wouldn't ``~clad::opts::tbr_analysis`` correspond to "all options enabled except for ``clad::opts::tbr_analysis``"? This would work if we took the bitwise & with this option. But we take the bitwise | which will only lead to all other options being enabled.

**Comment by vaithak** [comment]
> 
> Wouldn't `~clad::opts::tbr_analysis` correspond to "all options enabled except for `clad::opts::tbr_analysis`"? This would work if we took the bitwise & with this option. But we take the bitwise | which will only lead to all other options being enabled.

Ah, Good catch, thanks for pointing this out. 
Then, yeah, we should have two separate options only (enable, and disable).
Still, I will add some changes for the first benefit - type safety.

File: tools/ClangPlugin.cpp (hunk 1 above)
Comment (vaithak) [reply tools/ClangPlugin.cpp]: done 👍🏼

File: test/FirstDerivative/FunctionCalls.C [hunk 2]
```diff
@@ -175,6 +175,7 @@ int main () {
   clad::differentiate(test_6, "x");
   clad::differentiate(test_7, "i");
   clad::differentiate(test_8, "x");
-
+  clad::differentiate<clad::opts::enable_tbr>(test_8); // expected-error {{TBR analysis is not yet supported in forward mode.}}
```
[the remark this answers is not in the document]
Comment (vaithak) [reply test/FirstDerivative/FunctionCalls.C]: updated 👍🏼

File: tools/ClangPlugin.h [hunk 3]
```diff
@@ -74,22 +74,24 @@ namespace clad {
...
+          PrintNumDiffErrorInfo(false) {}
+
+    bool DumpSourceFn : 1;
+    bool DumpSourceFnAST : 1;
+    bool DumpDerivedFn : 1;
+    bool DumpDerivedAST : 1;
+    bool GenerateSourceFile : 1;
+    bool ValidateClangVersion : 1;
+    bool EnableTBRAnalysis : 1;
+    bool DisableTbrAnalysis : 1;
```
[the remark this answers is not in the document]
Comment (vaithak) [reply tools/ClangPlugin.h]: fixed 👍🏼

**Comment by vaithak** [comment]
> I am wondering if we could draw some inspiration from LLVM's BitmaskEnum?

My bad, I forgot that we are using the first few bits of this same mask for derivative order too. So, we can't have strict typing :(.

**Comment by vgvassilev** [comment]
@vaithak, @PetroZarytskyi is that ready to go?

**Comment by vaithak** [comment]
I have addressed the requested changes, @PetroZarytskyi if this looks good to you then we can go ahead.

**Comment by PetroZarytskyi** [comment]
The PR looks good to me. Thank you for the work!

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**